lldb/source/Plugins/LanguageRuntime/Swift/SwiftLanguageRuntimeDynamicTypeResolution.cpp

```diff
@@ -382,227 +383,289 @@ const CompilerType &SwiftLanguageRuntimeImpl::GetBoxMetadataType() {
-          } else {
-            stream << "\\0";
-            stream.PutHex8(c);
-          }
-        }
-        return stream.GetData();
-      };
-      LLDB_LOGV(log, "[MemoryReader] memory read returned data: \"{0}\"",
-                format_string(dest));
-      return true;
-    } else {
-      LLDB_LOGV(log, "[MemoryReader] memory read returned error: {0}",
-                error.AsCString());
-      return false;
-    }
+uint64_t LLDBMemoryReader::addModuleToAddressMap(ModuleSP module) {
+  // We set the highest bit to tell that the address is one mapped by us when
+  // doing memory reads later on.
+  uint64_t next_available_address = ((uint64_t)1 << 63);
+  if (!m_address_map.empty()) {
+    // We map the images contiguously one after the other, all with the highest
+    // bit set.
+    next_available_address = m_address_map.back().first + 1;
   }
+  auto *section_list = module->GetObjectFile()->GetSectionList();
+  uint64_t size = section_list->GetSectionAtIndex(0)->GetFileAddress();
+  for (auto &section : *section_list)
+    size += section->GetFileSize();
+  // Temporary hack with a size we're sure will fit the image
+  m_address_map.emplace_back(next_available_address + 0x400000000000, module);
+  return next_available_address;
+}
 
-  void pushLocalBuffer(uint64_t local_buffer, uint64_t local_buffer_size) {
-    lldbassert(!m_local_buffer);
-    m_local_buffer = local_buffer;
-    m_local_buffer_size = local_buffer_size;
-  }
+llvm::Optional<Address>
+LLDBMemoryReader::tryToConvertFromMappedAddressIntoRealAddress(
+    uint64_t address) const {
+  Log *log(GetLogIfAllCategoriesSet(LIBLLDB_LOG_TYPES));
 
-  void popLocalBuffer() {
-    lldbassert(m_local_buffer);
-    m_local_buffer.reset();
-    m_local_buffer_size = 0;
-  }
 
-private:
-  Process &m_process;
-  size_t m_max_read_amount;
+  // If the highest bit is set, this is an image we registered.
+  if (address >> 63) {
+    // Pair for the lower_bound function. Pair compares the first element first,
+    // so it should only ever need to compare the addresses and never the
+    // module.
+    auto comparison_pair = std::make_pair(address, ModuleSP());
+    auto pair_iterator = std::lower_bound(m_address_map.begin(),
+                                          m_address_map.end(), comparison_pair);
+    if (pair_iterator != m_address_map.end()) {
+      uint64_t masked_file_address = pair_iterator->first;
+      ModuleSP module = pair_iterator->second;
+      uint64_t file_address;
+      if (pair_iterator == m_address_map.begin())
+        // Since this is the first registered module,
+        // just unsetting the highest bit will give the real file address.
+        file_address = address & ~(((uint64_t)1) << 63);
+      else
+        // The start of the masked address is right after the previous
+        // registered address,
+        // so subtract it and 1 to get to the file address.
+        file_address = address - (pair_iterator - 1)->first - 1;
```

Comment:
Does this check that the address actually falls inside of the image that lower_bound() returned?

Reply:
Yes, by checking the iterator in the beginning: `pair_iterator != m_address_map.end`